### ISSUE TITLE
Add Action Sets to the Input Map

### ISSUE THREAD

**Issue by BastiaanOlij** [godotengine/godot-proposals#2273]
### Describe the project you are working on
Implementing OpenXR in Godot but this proposal applies to normal games just the same.

### Describe the problem or limitation you are having in your project
OpenXR has an input binding system that is very similar to Godot and the current approach we are taking is to add VR specific input events to the input map with which we can popular OpenXRs system or for other SDKs mimic the same behavior within our own solution. 

OpenXR introduces one feature that Godot doesn't have, the concept of Action Sets grouping a number of actions together. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Action sets group actions together which can be enabled/disabled based on game state. A good example of how this approach can benefit a game is looking at a game like FarCry or Grand Theft Auto where the player can control various vehicles with widely varying but often overlapping control schemes. 

One could define an action set for "first person control" that is active when the user is not in a vehicle and maps inputs to making the character walk around.
One could define an action set for "car control" that is active when the user is driving a car
One could define an action set for "helicopter control" that is active when the user is flying a helicopter

In VR action sets are often enabled/disabled depending on what tool the player holds in a given hand. 

By only having action sets enabled that are relevant to the current state of the game there is no need to ignore action events or deal with actions that are differently mapped depending on what vehicle is being controlled.

Multiple action sets can be active at a given time, one could isolate actions belong to menus or actions that should always be active regardless of game state in their own sets. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
I'll add a mockup screen image in due time but it involves adding the ability to create an action set within the input map pane in project settings and change the add action feature so you create the action within the action set.

An interface should be added to the Input Singleton that enabled/disables a given action set, the game loop that processes actions should only process actions for enabled sets.

The InputEventAction class should get a property that identifies the action set involved.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
One can get the same behavior by ignoring events or more carefully naming events so game code only reacts to the events intended for them. But as a game grows this can get messier. Solving it at the source prevents issues around this.

### Is there a reason why this should be core and not an add-on in the asset library?
This is an improvement on an already core part of the system.

**Comment by me2beats** [comment]
Do I get it right that `ImputMap` API should be changed then?
Something like 
`add_action(action, deadzone)` becomes `add_action(set, action, deadzone)`
(where set is set name: String), 
`erase_action(action)` becomes `erase_action(set, action)` etc

also some new methods should be added like
`Array get_active_sets()`,
`bool is_set_active(set)`,
`add_set(set)`
`void remove_set(set)`

Not familar with Godot 4 input system (is it different in this case?)
so I make these assumptions based on my knowledge of Godot 3 input map system.


Another question, how will this work with name collisions of different active sets?

If set `X` and set `Y` are active and both of them have a `move_forward` action,
then will both of them work at the same time?

**Comment by BastiaanOlij** [comment]
@me2beats yup, pretty much the way I see this working. I think a valid question is whether we want an `add_action_set(set)` to create a set first or whether the set is automatically created when the first action is added. 
I also think we should contemplate keeping `add_action(action, deadzone)` as a function and have set default to `default` or something so that we don't break peoples existing projects (maybe with a note that this version of `add_actions` is deprecated and will be removed in a future version).

Godot 4's input system as far as I know is currently unchanged from Godot 3.
For name collisions, you want to be able to re-use action names in different sets. OpenXR added a `priority` setting on the action set, the active action set with the highest priority will trigger the action.  I suggest we add the same mechanism.
If two action sets have the same priority it'll be whichever is higher in the action set list.

**Comment by groud** [comment]
I understand the idea, but I would prefer to keep the input mapping system as simple as possible. Input handling is one of the first things users have to deal with when they start creating a game, so I would prefer these things stays simple, without users requiring to deal with setting and enabling/disabling groups or things like that. We have to keep things simple to understand, and creating action sets might not be the simplest solutions.

I believe in the current InputMap's philosophy,  your initial problem should be solved by using a different action for each group of action. Taking the example of going left/right in two different gameplay, you would have those actions:
- drive_right
- drive_left
- walk_right
- walk_left
etc...

With this, "disabling" a set of action is then not even needed, because you check the correct action depending on your gameplay situation.

The only issue remaining here is about mapping. It gets a little bit more tedious as you have to synchronize the mapping of several actions. It's not that tedious to do via code, but the current input mapping interface does not work fine with that.

One idea would be to be able to create a hierarchy of actions. One action could "inherit" from another one, and if the parent action matches, then it means that the child action matches too. With the previous example, it would be something like:
- left -> you would set your mappings on this action
   - walk_left
   - drive_left
-right -> you would set your mappings on this action
   - walk_right
   - drive_right

It's not a lot more simpler but at least it does not involve disabling groups.

**Comment by 2plus2makes5** [comment]
I think a better way to do things would be to be able to save and load input maps, you could have one for walking, one for driving etc and then you load the one you need, this way they would also be way more reusable.

**Comment by BastiaanOlij** [comment]
Sorry for the late reaction.

@groud , thing is, it's pretty easy to ignore if you don't want to use it. Godot should have a default action set and that actionset should be enabled. If you just add all your actions to that default action set Godot works exactly as before.

But if you do wish to add additional action sets you can. 

The thing is, Godot is special I've found out in the fact that we actually offer a bindings solution out of the box. Many games just don't bother which is why Valve added bindings into Steam itself and thats followed through into OpenXR. With XR this is even more important, as more headset systems hit the market games only build to support say Vive Wands falter. 
The bindings we supply to OpenXR are simply defaults after which rebinding happens in the runtime so it is guaranteed people with headsets and controller invented after your game is released are able to play your game. 

Supporting this in full would be good but it isn't a show stopper if we decide not to. If we don't add action sets we'll just supply one big set to OpenXR and we work the way we have always done with the code in whatever part of the game that is running reacting to the correct actions.

That said, I actually like the approach and I think we can add it in without forcing it on anyone.

**Comment by BastiaanOlij** [comment]
> 
> 
> I think a better way to do things would be to be able to save and load input maps, you could have one for walking, one for driving etc and then you load the one you need, this way they would also be way more reusable.

Well you could already do that today, that would be pretty straight forward to build with Godots build in options. It wouldn't solve my issue though, with OpenXR you communicate your bindings at the start and seeing re-bindings now happen in the OpenXR runtime changing what actions are available would sever those bindings.

**Comment by BastiaanOlij** [comment]
I'm going to close this, there was too much difference between the action system in OpenXR and Godot so we did a implementation of the OpenXR action system as part of the OpenXR module.
